src/box/relay.cc

```diff
@@ -1042,11 +1042,17 @@ relay_subscribe(struct replica *replica, struct iostream *io, uint64_t sync,
 	assert(replica->anon || replica->id != REPLICA_ID_NIL);
 	struct relay *relay = replica->relay;
 	assert(relay->state != RELAY_FOLLOW);
+
 	/*
-	 * Register the replica with the garbage collector
-	 * unless it has already been registered by initial
-	 * join.
+	 * Reregister the replica with the garbage collector. To
+	 * correctly reregister the replica in gc, it's necessary
+	 * remove the old consumer.
 	 */
+	if (replica->gc != NULL) {
+		gc_consumer_unregister(replica->gc);
+		replica->gc = NULL;
+	}
+
```

Comment:
Let's simplify this a bit:
```diff
diff --git a/src/box/relay.cc b/src/box/relay.cc
index 87fd0a431..de0081367 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -1044,21 +1044,18 @@ relay_subscribe(struct replica *replica, struct iostream *io, uint64_t sync,
        assert(relay->state != RELAY_FOLLOW);
 
        /*
-        * Reregister the replica with the garbage collector. To
-        * correctly reregister the replica in gc, it's necessary
-        * remove the old consumer.
+        * Register the replica with the garbage collector.
+        * In case some of the replica's WAL files were deleted, it might
+        * subscribe with a smaller vclock than the master remembers, so
+        * recreate the gc consumer unconditionally to make sure it holds the
+        * correct vclock.
         */
-       if (replica->gc != NULL) {
-               gc_consumer_unregister(replica->gc);
-               replica->gc = NULL;
-       }
-
-       if (replica->gc == NULL && !replica->anon) {
+       if (!replica->anon) {
+               if (replica->gc != NULL) {
+                       gc_consumer_unregister(replica->gc);
+               }
                replica->gc = gc_consumer_register(replica_clock, "replica %s",
                                                   tt_uuid_str(&replica->uuid));
-               if (replica->gc == NULL)
-                       diag_raise();
-               gc_delay_unref();
        }
 
        if (replica_version_id < version_id(2, 6, 0) || replica->anon)
```

Reply:
`gc_consumer_register()` can return NULL. Should this part be removed?
```
    if (replica->gc == NULL)
            diag_raise();
    gc_delay_unref();
```

Reply:
> gc_consumer_register() can return NULL. Should this part be removed?

Good catch, I removed it by mistake. It should stay.